### PR TITLE
✨ performance optimization and runtime data-adjustmens with `hf_data` property

### PR DESCRIPTION
**Pull request by jonasvdd**
```python
import numpy as np
from plotly_resampler import FigureResampler
import plotly.graph_objects as go

x = np.arange(10_000_000)
noisy_sin = (3 + np.sin(x / 15_000) + np.random.randn(len(x)) / 10) * x / 1_000

fig = FigureResampler(go.Figure())
fig.add_trace(go.Scattergl(name="noisy sine", showlegend=True) , hf_x=x, hf_y=noisy_sin)

fig.update_layout(
    autosize=True,
    height=300,
    template="plotly_dark",
    legend=dict(x=0.1, y=1, orientation="h"),
    margin=dict(l=0, r=0, b=0, t=0, pad=0),
)
fig.show_dash(mode="inline")
```

To adjust data-properties one can now call:
```python
fig.hf_data[0]['y'] = -noisy_sin
```

`TODO`:
- [x] document this behavior
- [x] write tests
- [x] document new aggregators & add tests for them

---
This PR also includes a minor bug fix, allowing plotly-resampler to have more similar behavior as plotly. It now allows to pass (just as plotly) no data to scatter plots and only y data to scatter plots

### CODE REVIEW

File: plotly_resampler/aggregation/aggregation_interface.py [hunk 1]
```diff
@@ -60,24 +60,61 @@ def _supports_dtype(self, s: pd.Series):
...
+        # ----- divide and conquer heuristic to calculate the median diff ------
+        s_idx_diff = np.diff(s.index.values)  # fyi: s_idx_diff.shape === len(s) -1
+
+        # to do so - use a quantile-based (median) approach where we reshape the data
+        # into `n_blocks` blocks and calculate the min
+        n_blcks = 128
+        if s.shape[0] > n_blcks:
+            blck_size = (s_idx_diff.shape[0]) // n_blcks
+
+            # convert the index series index diff into a reshapable view (i.e., sid_v)
```
Comment (jvdd) [inline plotly_resampler/aggregation/aggregation_interface.py:74]:
```suggestion
            # convert the index series index diff into a reshaped view (i.e., sid_v)
```

File: plotly_resampler/aggregation/aggregation_interface.py [hunk 2]
```diff
@@ -60,24 +60,61 @@ def _supports_dtype(self, s: pd.Series):
...
+        # add None data-points in-between the gaps
+        if med_diff is not None:
+            df_gap_idx = s.index.values[1:][s_idx_diff > 2.1 * med_diff]
+            if len(df_gap_idx):
+                df_res_gap = pd.Series(
+                    index=df_gap_idx, data=None, name=s.name, copy=False
+                )
+
+                if isinstance(df_res_gap.index, pd.DatetimeIndex):
+                    print("converting to datetime index")
```
Comment (jvdd) [inline plotly_resampler/aggregation/aggregation_interface.py:97]: Would remove this print statement

File: plotly_resampler/figure_resampler.py [hunk 3]
```diff
@@ -748,7 +772,8 @@ def construct_update_data(self, relayout_data: dict) -> List[dict]:
             in each dict.
 
         """
-        current_graph = self.to_dict()
+        # see: https://github.com/plotly/plotly.py/blob/2e7f322c5ea4096ce6efe3b4b9a34d9647a8be9c/packages/python/plotly/plotly/basedatatypes.py#L3278
+        current_graph = {"data": self._data, "layout": self._layout}
```
Comment (jvdd) [inline plotly_resampler/figure_resampler.py:776]: Nice one! :fire:

File: plotly_resampler/figure_resampler.py [hunk 4]
```diff
@@ -417,6 +417,34 @@ def to_same_tz(
 
         return hf_series[to_same_tz(t_start) : to_same_tz(t_stop)]
 
+    @property
+    def hf_data(self):
+        return list(self._hf_data.values())
```
Comment (jvdd) [inline plotly_resampler/figure_resampler.py:422]: Calling `.values()` on a dict does not return a fixed order, right? Could this be a problem?

Comment (jvdd) [reply plotly_resampler/figure_resampler.py:422]: Oh wow, TIL "Dicts preserve insertion order in Python 3.7+"
https://stackoverflow.com/a/613218/12180712

File: tests/test_aggregators.py [hunk 5]
```diff
@@ -122,7 +122,10 @@ def test_lttb_categorical_time_data(cat_series):
 
     for n in np.random.randint(100, len(cat_series) / 3, 3):
         out = LTTB(interleave_gaps=True).aggregate(cat_series, n_out=n)
-        assert sum(out.notna()) == n
+        # LTTB uses the first and last value by default -> so it might add a none
+        # between these two peositions when not a lot of samples are chosen
```
Comment (jvdd) [inline tests/test_aggregators.py:126]:
```suggestion
        # between these two positions when not a lot of samples are chosen
```

File: tests/test_aggregators.py [hunk 6]
```diff
@@ -146,7 +153,7 @@ def test_lttb_bool_time_data(bool_series):
 
     for n in np.random.randint(100, len(bool_series) / 3, 3):
         out = LTTB(interleave_gaps=True).aggregate(bool_series, n_out=n)
-        assert sum(out.notna()) == n
+        assert sum(out.notna()) <= n
```
Comment (jvdd) [inline tests/test_aggregators.py:156]: Is this really desired behavior? Wouldn't we expect `==n`?

**Review by jvdd — APPROVED** [review]
LGTM, only documentation & tests of the property is needed!

**Comment by jonasvdd** [comment]
Thanks @mcourteaux for the devide-and conquer bin-size median approach!

File: plotly_resampler/aggregation/aggregation_interface.py [hunk 7]
```diff
@@ -60,24 +60,61 @@ def _supports_dtype(self, s: pd.Series):
...
+
+        # to do so - use a quantile-based (median) approach where we reshape the data
+        # into `n_blocks` blocks and calculate the min
+        n_blcks = 128
+        if s.shape[0] > n_blcks:
+            blck_size = (s_idx_diff.shape[0]) // n_blcks
+
+            # convert the index series index diff into a reshaped view (i.e., sid_v)
+            # Remark: we start from index "1" as index "0" contains a `NaN` value
+            sid_v: np.ndarray = s_idx_diff[: blck_size * n_blcks].reshape(n_blcks, -1)
```
Comment (mcourteaux) [inline plotly_resampler/aggregation/aggregation_interface.py:76]: NaN-bug coming from the `diff()` yielding NaN for the first element.

Comment (jonasvdd) [reply plotly_resampler/aggregation/aggregation_interface.py:76]: aah no my bad, s_idx_diff is now a numpy.ndarray and not the original series -> so there is no nan. The code comments are not up-to-data, hence the confusion

Comment (mcourteaux) [reply plotly_resampler/aggregation/aggregation_interface.py:76]: nice!

Comment (jonasvdd) [reply plotly_resampler/aggregation/aggregation_interface.py:76]: no no, that remark comment above is really confusing!

**Comment by jonasvdd** [comment]
# :fire: New aggregator function in town!

I hereby introduce the `MinMaxOverlapAggregator`, and let the screenshots speak for themselves

Note `s` in a `pd.Series`, with shape: `110,000,000`

⬇️ `LTTB` timings ⏱️ 
image


⬇️ `MinMaxOverlapAggregator` timings 🚀 
image


⬇️  (almost) no difference in aggregation output
image

**Comment by mcourteaux** [comment]
Beautiful! New paper?

**Comment by jonasvdd** [comment]
Ready for review @jvdd! 🔍

File: plotly_resampler/aggregation/aggregators.py [hunk 8]
```diff
@@ -63,39 +82,189 @@ def _aggregate(self, s: pd.Series, n_out: int) -> pd.Series:
...
+            + offset
+            + argmax_offset
+        )
+        # Sort the argmin & argmax (where we append the first and last index item)
+        # and then slice the original series on these indexes.
+        return s.iloc[np.unique(np.concatenate((argmin, argmax, [0, s.shape[0] - 1])))]
+
+
+class MinMaxAggregator(AbstractSeriesAggregator):
+    """Aggregation method which performs binned min-max aggregation over non-overlapping
```
Comment (jvdd) [inline plotly_resampler/aggregation/aggregators.py:147]: IMO these  are not non-overlapping but FULLY overlapping windows.
As the thing that overlaps is the min-window and max-window

**Review by jvdd — APPROVED** [review]
LGTM! Only doubting whether MinMaxAggregator is really non-overlapping (imo it is just FULLY overlapping).

File: plotly_resampler/aggregation/aggregators.py (hunk 8 above)
Comment (jonasvdd) [reply plotly_resampler/aggregation/aggregators.py:147]: Agree, good catch! 🧐

**Comment by mcourteaux** [comment]
Nice work! Can I pip install the git version to try it? :D Not sure how that works.

**Comment by jvdd** [comment]
> Nice work! Can I pip install the git version to try it? :D Not sure how that works.

Jup! Just released a new version on pypi (v 0.5.0)

https://pypi.org/project/plotly-resampler/

**Pull request merged**